Citation Popup: Search through the list of references using a given query

xUser5000 opened the pull request:
## What has been done
- Pass reference data as JSON to the citation popup.
- Integrate [AutoComplete.js](https://tarekraafat.github.io/autoComplete.js/): I found it to be more suitable for the task than [Typeahead.js](https://twitter.github.io/typeahead.js/). In addition, it has a lot of customizations with zero dependencies.
- Enable selecting references from search results.
- Enable removing references from the selected area.
- Prevent showing already-selected references in the search results.
- Return array of selected references and insert them into the note body.

## Demo
![Demo](https://i.ibb.co/kHLKgmz/DEMO-SEARCH-CITATION-POPUP.gif)

## Bugs
The following bugs still need to be solved:
- Inserted references often don't go to the desired location in the note body; some of them go to the desired location and the rest is inserted at the top of the note body.
- Enter key does the functionality of the "OK" button; It should select the highlighted reference in the search results.

laurent22 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,90 @@
+    document.getElementById(refId).remove();
+    if (selectedRefs.size === 0) {
+        selectedRefsView.textContent = "Select some references to be added to the current note";
+    }
```
Manipulating DOM nodes like this and trying to keep that in sync with the state is going to be unmaintainable over the long term. Is it not possible to use a lib to do that for you?

laurent22 reviewed (changes requested):
Nice to see you are making good progresses. While you develop, could you also test that your code works with large BibTex files? I would suggest finding a 1GB file and see how it performs as that would be a good way to stress test your code. It's better to do this from time to time as you make progresses, because it will inform how you should develop certain parts of the plugin.

xUser5000 commented on src/add-bibtex-reference.command.ts:
```diff
@@ -40,19 +40,21 @@ export async function registerAddBibTexReferenceCommand () {
+                if (selectedRefsIDs.length === 0) return;
+
+                // Insert the selected references into the note content
+                for (let i = 0; i < selectedRefsIDs.length; i++) {
+                    const selectedReference = DataStore.getReferenceById(selectedRefsIDs[i]);
+                    await joplin.commands.execute(
+                        "insertText",
+                        formatReference(selectedReference)
+                    );
+                }
```
Wow, that's a very good idea

xUser5000 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,90 @@
+const output = document.getElementById("output");
+
+const refs = JSON.parse(inputRefsView.textContent);
+const refsMap = new Map();
+refs.forEach(ref => { refsMap.set(ref["id"], ref); });
```
My point is that it allows for fast lookups. For example, when I want to get the reference represented by a given id, I look it up from the map.

laurent22 replied:
What's the difference between using an object, and using a map in terms of performance?

xUser5000 commented:
- I looked into different libraries and frameworks (like [Svelte](https://svelte.dev/), [Stimulus](https://stimulus.hotwire.dev/), [Bind-dom](https://github.com/shystruk/bind-dom) ) that do state-based UI rendering but I felt like they are overkill when compared to the task at hand. Based on that, I decided to write it myself. Performance-wise, it is not that great, but I think it is more extendable and clean than before. I don't know if this is the best thing to do.
- In addition, I solved the bug related to the wrong cursor position.
- There was also another bug that I found when stress-testing the code with ~1200 BibTeX references; the bug was that the parsing library does not guarantee that there will be a property `issued.["date-parts"]`. Accordingly, when calling the `getDate()` function on a reference that has the property undefined, it throws an error saying "Cannot read properly 0 of undefined". I solved this bug by simply checking of the desired property is defined or not.

laurent22 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,107 @@
+        refs
+            .map(refId => refsMap.get(refId))           // id => reference
+            .map(ref => (`
+                <li id="${ref["id"]}">
```
ID needs to be escaped (even if now you know it's not needed, it's always safer to do so, for example if later on the ID format is changed).

xUser5000 replied:
Is it necessary? I escape the JSON data before passing it to the view. Isn't that enough?

laurent22 replied:
Yes, escaping variables is not something optional to be added after later on. And I don't see what you mean by escaping the JSON data before passing it to the view

- First, if these properties are already escaped, then they are not named correctly. It's not just an "id" it's an "idHtml" - it's HTML code basically so it can't be used in an SQL query for example or other context that would need different escaping (see [Making Wrong Code Look Wrong](https://www.joelonsoftware.com/2005/05/11/making-wrong-code-look-wrong/) for good ideas on how to name variables)

- Secondly, you should only escape data at the point where it is needed. Pre-escaping data is a recipe for accidentally double-escaping, and often these issues are not obvious during development, only when people start importing all kind of unexpected data.

Somehow we keep going back to these escaping issues, so please try to remember to always escape data that you insert into strings. Perhaps do some reading about the topic to better understand how important it is. But to sum up:

- If you insert the data into HTML, use HTML entities encoding

- If you insert into a URL, use encodeURIComponent or encodeURI (you have the article where they explain the difference)

laurent22 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,107 @@
+            .map(refId => refsMap.get(refId))           // id => reference
+            .map(ref => (`
+                <li id="${ref["id"]}">
+                    <span class="title">${ref["title"]}</span>
```
Title needs to be escaped.

laurent22 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,107 @@
+ * @param {Reference[]} refs
+ * @returns string
+ */
+function template (refs = []) {
```
That's a good idea to do all the rendering in one place like this. Ideally a lib like React would be better but I can agree it's a bit overkill at this stage.

xUser5000 commented:
Regarding the difference between maps and plain js objects, it turns out that objects are much faster than maps in initialization and assigning values to keys. Objects are not as flexible as maps, but, in this situation, I don't that much flexibility (e.g getSize(), delete()); I only need to add key-value pairs and then look up the reference that corresponds to a given key. In conclusion, plain objects are better suited for this task.

laurent22 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,107 @@
+                // Modify Results Item Content
+                item.innerHTML = `
+                    <span style="text-overflow: ellipsis; white-space: nowrap; overflow: hidden;">
+                        ${data.match}
```
escape

xUser5000 commented:
I've replaced the map with a plain js object and fixed minor UI bugs.
Now my question is: How to implement escaping in `view.js`? Do I download a new library just for that or? I also want to know whether or not this is necessary because I already escape the JSON object passed to the view.

laurent22 commented:
> Now my question is: How to implement escaping in view.js?

That's all you need:

`<li id="${encode(ref["id"])}">`

xUser5000 commented:
> That's all you need:
> 
> `<li id="${encode(ref["id"])}">`

But I don't have access to the library from thee view, so I will need to download a new library specifically for this.

xUser5000 commented:
> That's all you need:
> 
> `<li id="${encode(ref["id"])}">`

But I don't have access to `html-entites` from the view, so there are three options at hand:
- Download a new library specifically for this.
- Write a basic `encode()` function from scratch.
- Let the DOM handle the escaping for me. For example,
```js
const li = document.createElement("li");
li.textContent = ref["id"];   // Auto-escaping
```

laurent22 commented:
> But I don't have access to html-entites from the view, so there are three options at hand:

All these options are fine so it's up to you.

xUser5000 commented:
- Not encoding the JSON string passed by `citation-popup/index.ts` was causing a bug. For instance, if I try to include a title like "This is an <art", it gets stringified to JSON peacefully but it breaks the HTML in the view. So, I had to encode the JSON string before passing it, then decode it when receiving it in the view.
- I used [he.js](https://github.com/mathiasbynens/he) to encode and decode data inside the view. With that said, I encountered a bug that was mentioned [here](https://discourse.joplinapp.org/t/dialog-plugin-communication/17956/7) by Ahmed and solved it the same way he did (by calling `setTimeout()` for a fraction of a second).
- In addition to the above, I did extra minor things like refactoring loading assets into a separate function in `citation-popup/index.ts` and made the search field empty its contents after the user adds a reference to the selected area.

laurent22 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,125 @@
+                    // Modify Results Item Content
+                    item.innerHTML = `
+                        <span style="text-overflow: ellipsis; white-space: nowrap; overflow: hidden;">
+                            ${ data.match }
```
Shouldn't you escape this too?

xUser5000 replied:
OH, I forgot to mention that `data.match` is a DOM element that's responsible for highlighting the parts of the title that match the search query, so we can assume that it does escaping by default (I guess because it uses `textContent` internally). Also, it didn't break when I used the title 'This is an <art".

laurent22 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,125 @@
+function main () {
+
+    /* State */
+    const refs = JSON.parse( he.decode(inputRefsView.textContent) );
```
I don't think you need to decode here - textContent is plain text, not HTML.

laurent22 commented on src/ui/citation-popup/index.ts:
```diff
@@ -12,46 +12,55 @@ let popupHandle: string = "";
     );
-
-    const refs: Reference[] = DataStore.getAllReferences();
     html = html.replace("<!-- content -->", fromRefsToHTML(refs));
```
By the way, did you check how it works with a 1GB file?

xUser5000 commented on src/ui/citation-popup/view.js:
```diff
@@ -1,22 +1,125 @@
+waits until all the other scripts get loaded by Joplin,
+and then starts doing its job
+*/
+setTimeout(main, 5, 10);
```
Got it

xUser5000 commented:
# Whats has been done
OK, I've removed the use of the dictionary, replaced `setInterval()` with `setTimeout()`, and removed the redundant escaping in `view.js`.

# Stress test
In addition to the above tasks, I stress tested the code with large `.bib` files. I already did a stress test with a ~300KB file, but this time I tried much larger files.

## Method
I couldn't find a bibtex file with 1GB of data so I downloaded [this](https://shelah.logic.at/v1/listb.bib) file. After that, I wrote a [script](https://ideone.com/G53WKF) that makes another copy of the file and then starts duplicating its contents over and over again until it reaches a certain size that I specify.

## Results
After adjusting the size parameter and observing what happens to the plugin when I execute the main command, I came up the following conclusions:
- If the size is above `512MB` (about equivalent to `536870912 Bytes`, which is about equivalent to `0x1FFFFFE8 Bytes` in Hexadecimal), nothing happens for one or two seconds then the below alert appears:
![Cannot create a string longer than 0x1FFFFFE8 characters](https://i.ibb.co/CWM6tQG/Screenshot-from-2021-07-03-00-01-31.png)
- If the size is below `512MB`, the call stack will not afford that much data and will throw an error like this:
![Maximum call stack size exceeded](https://i.ibb.co/qn1ttVm/Screenshot-from-2021-07-03-00-07-32.png)
- If the size `<= 12MB`, the plugin works fine and there are no issues at all.

laurent22 commented:
Looks good now, let's merge. Regarding the stress test, could you put this into a new issue instead please? (otherwise the info will be lost once the PR is closed)
